micro_manager/micro_manager.py

```diff
@@ -606,20 +702,87 @@ def _solve_micro_simulations_with_adaptivity(
+                        "See next entry for error message.".format(
+                            self._mesh_vertex_coords[active_id]
+                        )
+                    )
+                    self._logger.error(error_message)
+                    self._crashed_sims[active_id] = True
+
+        # Interpolate result for crashed simulation
+        if self._crashed_sims.count(True) > 0:
+            unset_sims = []
+            for active_id in active_sim_ids:
+                if micro_sims_output[active_id] is None:
+                    unset_sims.append(active_id)
+
+            # Interpolate output for crashed simulations
+            for unset_sim in unset_sims:
+                self._logger.info(
+                    "Interpolating output for crashed simulation at macro vertex {}.".format(
+                        self._mesh_vertex_coords[unset_sim]
+                    )
+                )
+                coord = []
+                interpol_values = []
+                output_interpol = dict()
+                # Collect neighbor vertices for interpolation
+                for neighbor in self._neighbor_list[unset_sim]:
+                    if not self._crashed_sims[neighbor] and neighbor in active_sim_ids:
+                        coord.append(self._mesh_vertex_coords[neighbor].copy())
+                        interpol_values.append(micro_sims_output[neighbor].copy())
+                        interpol_values[-1].pop("micro_sim_time", None)
+                        interpol_values[-1].pop("active_state", None)
+                        interpol_values[-1].pop("active_steps", None)
+                if len(coord) == 0:
+                    self._logger.error(
+                        "No neighbors available for interpolation at macro vertex {}.".format(
+                            self._mesh_vertex_coords[unset_sim]
+                        )
+                    )
+                else:
+                    # Interpolate output for each parameter
+                    for key in interpol_values[0].keys():
+                        key_values = []
+                        for elems in range(len(interpol_values)):
+                            key_values.append(interpol_values[elems][key])
+                        output_interpol[
+                            key
+                        ] = self._interpolation.inv_dist_weighted_interp(
+                            coord, self._mesh_vertex_coords[unset_sim], key_values
+                        )
+                    micro_sims_output[unset_sim] = output_interpol
+                    if self._is_micro_solve_time_required:
+                        micro_sims_output[unset_sim]["micro_sim_time"] = 0
+                    micro_sims_output[unset_sim]["active_state"] = 1
+                    micro_sims_output[unset_sim][
+                        "active_steps"
+                    ] = self._micro_sims_active_steps[unset_sim]
```

Comment:
We can probably move this into a function and call it once to keep the code clean. Right now we have too many if condition and loop going around, which makes it extremely hard to read code. In general up to 3 levels of loops are fine, more are a problem.